### ISSUE TITLE
[FIX-008] Re-elect the best completed run when the current best is overwritten

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#207]
**Category:** Resume and artifact integrity · **Priority:** P1 · **Size:** M

**Evidence status:** Reproduced defect.

## Outcome and scope

When destructive overwrite invalidates the best run, elect the best surviving committed candidate instead of comparing the replacement with itself. Keep the existing metric fallback/comparison order and atomic pointer formats. Scan on invalidation; a new database, general experiment index or named-monitor feature is out of scope.

## Current behavior / evidence

**Current behavior:** Overwriting the current best deletes its old artifacts and reuses its run ID. After replacement, `_update_best_pointer` follows `runs/best` to that same ID and compares the new run to itself. A better surviving run is never considered.

**Concrete reproduction:** Save completed run A with error 0.1 and run B with error 0.2; each has an IDP, LAST, BEST, and run YAML/CSV. A owns `runs/best`. Enter `A.writable_lease(overwrite=True)` and replace A's history/LAST/BEST with error 0.9. After A's save, `runs/best` still points at A although B is better. The updated probe verifies both `NNRun.load(A).idps` and `NNRun.load(B).idps` contain one row, so the comparison is between completed valid artifacts, not incomplete directories.

## Repository map

- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L374) — Overwrite deletes the old owner while retaining its content-addressed ID.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L390) — Stable per-run lease spans admission through final persistence.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L488) — Current cross-run comparison follows the old ID into the replacement.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L102) — Atomic symlink/pointer-file publication; `_read_best_pointer` handles both.
- [`src/nnx/nn/params/nn_run.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py#L508) — Use committed history reconciliation, not raw CSV tails.
- [`tests/test_train_integration.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_train_integration.py#L1245) — Existing overwrite lifecycle coverage.
- [`tests/test_pass2_r_series.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_pass2_r_series.py#L46) — Atomic persistence/failure fixtures.

## Implementation approach

1. Under the established ownership locks, detect whether the run about to be overwritten is the current best. Invalidate/re-elect that pointer before exposing the emptied replacement as a valid candidate. If there is no valid survivor, represent no best rather than pointing to an incomplete run. A later completed replacement participates normally.
2. Scan only actual run directories. Exclude `best`, `.leases`, reservations, malformed/unloadable metadata and missing/corrupt BEST checkpoints. Use NNRun.load’s LAST/history protocol to reject a BEST/history epoch beyond the committed boundary. Share candidate eligibility with the ordinary `_update_best_pointer` path: even its absent/dangling-pointer branch must validate the current run before publication. Saving incomplete metadata/history remains allowed, but cannot select a run without an eligible BEST. Apply LAST/history checks when the history protocol is present; keep legacy artifacts without that marker on their documented compatibility path. Distinguish invalid/missing checkpoints from valid committed all-None metrics, whose no-signal behavior is retained by FIX-009. Select with `_best_err` and existing strict-improvement behavior; retain the current winner on an equal eligible score where possible and document deterministic tie handling during a fresh election.
3. Keep `.best.lock` as the pointer election/publication serialization boundary. Document acquisition order relative to lease, admission, run and checkpoint locks. Do not hold the global best lock while blocking on another live run’s lease or run lock; collect/validate committed snapshots with a recheck strategy that avoids deadlock with `_save_locked` → `_update_best_pointer`.
4. Use `_point_best` for publication and preserve relative symlink portability and POINTER.txt fallback. Failed overwrite and interrupted replacement must never advertise a reservation. Keep FIX-009’s finite-score handling compatible without making this independent repair wait for new monitor features.

## Acceptance criteria

- [ ] The A=0.1/B=0.2/A-overwritten=0.9 example repoints to B.
- [ ] Overwriting A with 0.05 keeps A best; overwriting a non-best run does not unnecessarily disturb the pointer.
- [ ] Only valid committed run artifacts participate: ignore lease directories, incomplete reservations, malformed/absent BEST checkpoints, and history beyond LAST when the history protocol applies.
- [ ] A failed overwrite does not advertise an incomplete run as best; pointer-file fallback and symlink mode agree.
- [ ] Re-election preserves existing live-run ownership and atomic pointer replacement, with bounded concurrency coverage when the environment permits it.

## Validation and completion

Proposed regression fixture using existing APIs, with all model/run setup explicit (run inside pytest `tmp_path` via `monkeypatch.chdir(tmp_path)`):

```python
from nnx import (Checkpoints, Devices, Losses, Nets, NNCheckpoint,
                 NNEvaluationDataPoint, NNIterationDataPoint, NNModel,
                 NNModelParams, NNParams, NNRun, NNTrainParams)
from nnx.nn.params.nn_run import _read_best_pointer

def test_best_overwrite_regression(tmp_path, monkeypatch):
    monkeypatch.chdir(tmp_path)
    model = NNModel(net_params=NNParams(input_dim=2, output_dim=2, dropout_prob=0),
                    params=NNModelParams(net=Nets.FEED_FWD, device=Devices.CPU,
                                         loss=Losses.CROSS_ENTROPY))
    def completed(salt, error):
        edp = NNEvaluationDataPoint(f1=1-error, recall=1-error, accuracy=1-error,
                                   precision=1-error, loss=1., error=error)
        idp = NNIterationDataPoint(lr=.01, iter_idx=0, epoch_idx=0, batch_idx=0,
                                   train_edp=edp)
        run = NNRun(net=model.net_params, model=model.params,
                    train=NNTrainParams(n_epochs=1), salt=salt, idps=[idp])
        checkpoint = NNCheckpoint(net_params=model.net_params,
                                  model_params=model.params,
                                  net_state=model.net.state_dict(), idp=idp)
        for tag in (Checkpoints.LAST, Checkpoints.BEST):
            checkpoint.save(run=run.id, type=tag)
        return run.save()
    a, b = completed("A", .1), completed("B", .2)
    with a.writable_lease(overwrite=True):
        completed("A", .9)
    assert _read_best_pointer("runs/best") == b.id
```

Add `test_best_re_elected_after_best_overwrite`, `test_failed_best_overwrite_leaves_valid_survivor`, and `test_re_election_ignores_uncommitted_candidates`. Run both symlink and forced fallback modes. Test replacement scores `.05` and `.9`, non-best overwrite, empty/incomplete/malformed candidates and speculative history after LAST. Add `test_incomplete_replacement_save_cannot_recreate_best`: overwrite the sole winner, let re-election leave no pointer, call the replacement's ordinary `save()` before BEST exists, then inject failure; no best pointer may appear. Repeat with a valid survivor B and require B to remain selected. Reuse bounded lease-concurrency fixtures; if OS shared-memory restrictions prevent process execution, report the missing gate and run it on supported CI rather than declaring it covered.

From the repository root, use `uv sync --frozen --all-extras` for the CI-equivalent environment. Existing focused coverage (files already exist):

```bash
uv run --frozen pytest tests/test_train_integration.py tests/test_pass2_r_series.py
```

Add the proposed regression cases above to those existing test files, then rerun the same command. Finish with:

```bash
uv run --frozen ruff check src/ tests/ examples/ scripts/
uv run --frozen ruff format --check src/ tests/ examples/ scripts/
uv run --frozen pyright --warnings
```

No new dependency is needed for the implementation described. Preserve Python 3.10 compatibility and default-field omission in existing serialization. These commands are completion requirements for implementation; this specification review does not claim they were run or passed.

## Cascading changes and affected consumers

- **Both training entry points.** Trace overwrite through [NNRun.writable_lease/ensure_writable/save](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py) as reached by [NNModel.train](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py) and [Trainer.train](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/trainer.py). Recovery cannot depend only on a direct test calling `_update_best_pointer`: deleting the old winner before a replacement fit, failing before its first committed checkpoint, and successfully committing a worse replacement must all yield a valid surviving winner. Preserve lock ordering, admission ownership and atomic symlink/pointer-file replacement while invoking the common re-election logic.

- **Committed artifact readers.** Candidate history must use [NNRun.load](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/params/nn_run.py) so rows beyond committed LAST are ignored under the history protocol; a directory listing and a raw CSV minimum are insufficient. Validate BEST as parseable, committed checkpoint metadata before advertising its run, and exclude `.leases`, incomplete reservations and malformed candidates without abandoning valid survivors. Election must not construct arbitrary models or invoke registered factories. As an integration oracle on the chosen built-in fixture, follow the selected pointer through checkpoint loading and NNModel reconstruction and compare its predictions; reconstruction is not a universal eligibility requirement. Keep legacy finite lower-is-better ranking and deterministic tie policy; non-finite hardening remains coordinated with [FIX-009](https://github.com/thekaveh/NNx/issues/208).

- **Bounded recovery demonstration.** Add an explicit temporary-root recovery helper to [Example 02](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/02_resume_training.py), separate from its ordinary resume main. Create two tiny committed runs, deliberately overwrite the old winner with a worse result, and assert the surviving winner can be loaded for inference; also exercise a failure before replacement commit in tests. Document what overwrite replaces, when a run becomes eligible again and how readers observe a missing winner during recovery. Keep such destructive teaching fixtures confined to their fresh temporary root rather than enabling overwrite in the normal user workflow.

### Samples and executable coverage

- [`examples/02_resume_training.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/02_resume_training.py): Add an opt-in temporary-root helper with two tiny committed runs and a worse overwrite; load the re-elected BEST for inference and verify its run ID and predictions. **Verify:** Proposed execution node: uv run --frozen pytest tests/test_examples_smoke.py -k overwrite_best_recovery. CPU, offline, temporary working directory; call the bounded helper rather than the full training main. Retain import-without-main coverage and add the callable to explicit execution coverage; import success alone is insufficient.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/concepts.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/concepts.md) | `generated/site/Concepts.md` | `generated/wiki/Concepts.md` | Describe overwrite-triggered best re-election, committed candidate eligibility, failure recovery, and equivalent symlink/pointer-file semantics without changing legacy finite ranking. |
| [`docs/quickstart.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/quickstart.md) | `generated/site/Quickstart.md` | `generated/wiki/Quickstart.md` | Explain that deliberate overwrite replaces one run and re-elects the remaining best; show recovery only in an explicitly isolated temporary-root example. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Both NNModel.train and Trainer.train exercise a real overwrite of the current winner and leave the pointer resolving to a valid surviving run after a worse replacement or pre-commit failure.
- [ ] A re-elected built-in fixture loads through NNRun/NNCheckpoint and reconstructs NNModel with expected saved predictions; election itself inspects committed, parseable metadata without invoking arbitrary model factories, and ignores uncommitted CSV rows.
- [ ] The recovery path does not release another run’s live lease or expose a partially written pointer, and candidate validation errors do not suppress unrelated valid survivors.
- [ ] The bounded Example 02 overwrite helper executes only inside a fresh temporary run root and proves that the re-elected artifact remains usable for inference.
- [ ] The issue-specific changes in `docs/concepts.md`, `docs/quickstart.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** Propagation belongs to this fix and existing APIs. Related tickets coordinate shared contracts/files; they are not prerequisites for the bounded repaired path.

Additional nonblocking consumers/coordination: [FIX-007](https://github.com/thekaveh/NNx/issues/206), [FIX-009](https://github.com/thekaveh/NNx/issues/208), [FEAT-003](https://github.com/thekaveh/NNx/issues/213).

## Delivery order and dependencies

- **Execution order:** 08 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-009](https://github.com/thekaveh/NNx/issues/208), [FEAT-005](https://github.com/thekaveh/NNx/issues/215). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. No feature dependency; shares finite-score handling with FIX-009.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved.

**Root cause:** overwriting a run deletes its artifacts and reuses its ID; when that run owned `runs/best`, `_update_best_pointer` followed the pointer to the same ID and compared the replacement with itself, so a better surviving run was never considered and a worse (or emptied) replacement stayed advertised.

**What changed**
- `NNRun.ensure_writable(overwrite=True)` elects the best surviving committed run under `.best.lock` *before* deleting the winner's artifacts; with no eligible survivor the pointer is removed (documented "no best" state). The pointer never dangles and an emptied replacement is never advertised.
- Eligibility (`_committed_best_checkpoint`): real run dir with `run.yaml`, history loadable via `NNRun.load` (rows beyond committed LAST reconciled), parseable BEST, and under the history protocol a BEST epoch ≤ committed LAST; `best`, `.leases`, bare reservations and malformed artifacts are skipped without abandoning valid survivors. Election keeps the incumbent on ties, then the lexicographically smallest ID.
- `_update_best_pointer` validates the saving run before publication (an incomplete replacement's history save cannot recreate the pointer); absent/dangling/self-owned pointer → full election; otherwise the existing strict lower-is-better comparison. Publication still uses `_point_best` (symlink + `POINTER.txt` fallback). Lock order documented: lease → admission → best (overwrite), lease → run → best (save); election reads other runs locklessly.
- Example 02: bounded `overwrite_best_recovery()` (temporary root only; `tests/test_examples_smoke.py -k overwrite_best_recovery`). Docs: `concepts.md` §4.1, examples catalog, CHANGELOG, API reference.

**PRs:** #273 (develop, merged), #274 (main, merged).

**Checks run (both branches):**
- `uv run --frozen pytest -q` → 1605 passed, 1 skipped, 2 deselected (the two libcairo-dependent `test_docs_projection` tests — pre-existing local-env failures, unrelated; CI runs them and passed). New: `test_best_re_elected_after_best_overwrite` (symlink + pointer-file modes), `test_failed_best_overwrite_leaves_valid_survivor`, `test_incomplete_replacement_save_cannot_recreate_best`, `test_re_election_ignores_uncommitted_candidates`, `test_train_overwrite_failure_before_commit_re_elects_survivor` (real training entry point + `NNModel.from_checkpoint` prediction oracle).
- `ruff check` → All checks passed; `ruff format --check` → all formatted; `pyright --warnings` → 0 errors, 0 warnings
- docs chain (`build_api_reference` / `build_docs` / `mkdocs build --strict` / `build_wiki --check` / projection tests) → pass
- CI: lint-and-test on Python 3.10–3.14, pip-audit, CodeQL, docs build all green on #273 and #274.

**Not verifiable locally:** none beyond the pre-existing thread/process lease-concurrency suite, which passed locally and in CI; live Pages/wiki publication is left to the `docs.yml` main-push job.
